src/realm/object-store/sync/sync_session.cpp

```diff
@@ -326,53 +327,39 @@ SyncSession::handle_refresh(const std::shared_ptr<SyncSession>& session, bool re
 
 SyncSession::SyncSession(SyncClient& client, std::shared_ptr<DB> db, const RealmConfig& config,
                          SyncManager* sync_manager)
-    : m_config(config)
-    , m_db(std::move(db))
-    , m_flx_subscription_store([this](bool use_flx_sync) -> std::shared_ptr<sync::SubscriptionStore> {
-        if (!use_flx_sync) {
-            return nullptr;
-        }
-
-        return sync::SubscriptionStore::create(m_db, [this](int64_t new_version) {
-            util::CheckedLockGuard lk(m_state_mutex);
-            if (m_state != State::Active && m_state != State::WaitingForAccessToken) {
-                return;
-            }
-            // There may be no session yet (i.e., waiting to refresh the access token).
-            if (m_session) {
-                m_session->on_new_flx_sync_subscription(new_version);
-            }
-        });
-    }(m_config.sync_config&& m_config.sync_config->flx_sync_requested))
+    : m_db{std::move(db)}
+    , m_migration_store{sync::MigrationStore::create(m_db,
+                                                     [this](sync::MigrationStore::MigrationState) {
+                                                         // Migration state changed - Restart the session with the
+                                                         // updated configuration update_configuration() will
+                                                         // create/delete the subscription store, if switching between
+                                                         // pbs and flx
+                                                         auto sync_config = m_migration_store->convert_sync_config(
+                                                             m_original_sync_config);
+                                                         // Update the SyncSession's sync_config and restart the
+                                                         // session connection
+                                                         update_configuration(sync_config);
+                                                     })}
+    , m_original_sync_config{config.sync_config}
```

Comment:
I think for safety's sake we need to initialize m_original_sync_config before we create the migration store. But I'm also not sure why we need the original config?

Reply:
The `m_original_sync_config` is used to restore the PBS config when the migration is rolled back. I could just save the partition string, since that's really all that is needed.

Reply:
I agree with Jonathan. At the same time, I discussed with Michael that it's probably not needed to call update_configuration (and hence the need to convert the sync config) in the callback, so it'll be safe. Side note: I am not sure the MigrationStore's callback is really needed, so I may remove it altogether.